src/Lemon/Containers/Converters/Ncch2Binary.cs

```diff
@@ -0,0 +1,237 @@
+            WriteOffsetSizeAndData(writer, root.Children["system"], source.Header.SystemOffset, source.Header.SystemHashSize, true);
+
+            writer.Stream.Position += 0x4; // Reserved
+            WriteOffsetSizeAndData(writer, root.Children["rom"], source.Header.RomOffset, source.Header.RomHashSize, true);
+
+            writer.Stream.Position += 0x4; // Reserved
+
+            WriteSHA256(writer, root.Children["system"], source.Header.SystemHashSize, true);
+            WriteSHA256(writer, root.Children["rom"], source.Header.RomHashSize, true);
+
+            // Write the full size of the binary (in units) in 0x104
+            writer.Stream.Position = 0x104;
+            writer.Write((int)writer.Stream.Length / NcchHeader.Unit);
+
+            return binary;
+        }
+
+        void WriteFile(DataWriter writer, Node file, bool isRequired = true)
+        {
+            if (file == null && isRequired) {
+                throw new System.IO.FileNotFoundException("Missing NCCH file");
+            } else if (file == null) {
+                return;
+            }
+
+            if (file.Format is not IBinary) {
+                throw new FormatException("Cannot write file as it is not binary");
+            }
+
+            file.Stream.WriteTo(writer.Stream);
+        }
+
+        void WriteOffsetSizeAndData(DataWriter writer, Node file, int offsetInUnits, int hashRegion = 0, bool hasHashRegion = false)
+        {
+            if (file != null) {
+                long position = writer.Stream.Position;
+                int offset = offsetInUnits * NcchHeader.Unit;
+                int fileLength = (int)file.Stream.Length;
+                int fileLengthInUnits = fileLength / NcchHeader.Unit;
+
+                bool usedChunk = false;
+                var fileRange = Enumerable.Range(offset, fileLength);
+                foreach (var usedUnit in usedUnits) {
+                    var usedRange = Enumerable.Range(usedUnit.Key, usedUnit.Value);
+
+                    if (fileRange.Intersect(usedRange).Any()) {
+                        usedChunk = true;
+                    }
+                }
+
+                if (writer.Stream.Length > offset && usedChunk) {
+                    offset = (int)Math.Ceiling((double)writer.Stream.Length / NcchHeader.Unit) * NcchHeader.Unit;
+                    offsetInUnits = offset / NcchHeader.Unit;
+                }
+
+                if (writer.Stream.Length < offset) {
+                    long tempPosition = writer.Stream.Position;
+
+                    writer.Stream.Position = writer.Stream.Length;
+                    writer.Write(new byte[offset - writer.Stream.Length]);
+
+                    writer.Stream.Position = tempPosition;
+                }
+
+                writer.Write(offsetInUnits);
+                writer.Write(fileLengthInUnits);
+
+                writer.Stream.Position = offset;
+
+                file.Stream.WriteTo(writer.Stream);
```

Comment:
After writing the file, we need to add the padding to the unit: `writer.WritePadding(0, NcchHeader.Unit);`

Reply:
The file padding is still missing, right?

Reply:
I don't know how this works, but this isn't the content of the latest commit, but in there the file length is padded, so I think that should do it

Reply:
Yes, the ` fileLength` is padded, so we write the correct length value.
But we are not writing the actual padding bytes (`writer.WritePadding`). So the offset we calculate won't be right.

Reply:
Just realized that and edited the comment, now it should be padded.